File: backend/services/grammarService.py
#!/usr/bin/env python3
"""
Grammar Service for AdmitAI Korea
Uses LanguageTool for free, powerful grammar checking
"""

import json
import sys
from flask import Flask, request, jsonify
from flask_cors import CORS
from language_tool_python import LanguageTool
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = Flask(__name__)
CORS(app)

# Initialize LanguageTool
try:
    tool = LanguageTool('en-US')
    logger.info("LanguageTool initialized successfully")
except Exception as e:
    logger.error(f"Failed to initialize LanguageTool: {e}")
    tool = None

class GrammarAnalyzer:
    def __init__(self):
        self.tool = tool
        self.error_categories = {
            'GRAMMAR': 'Grammar errors',
            'SPELLING': 'Spelling mistakes',
            'STYLE': 'Style issues',
            'PUNCTUATION': 'Punctuation errors',
            'TYPOS': 'Typographical errors'
        }
    
    def analyze_text(self, text):
        """Analyze text for grammar, spelling, and style errors"""
        if not self.tool:
            return {
                'error': 'LanguageTool not available',
                'score': 0,
                'issues': []
            }
        
        try:
            # Get matches from LanguageTool
            matches = self.tool.check(text)
            
            # Process matches
            issues = []
            total_errors = len(matches)
            
            for match in matches:
                issue = {
                    'type': match.ruleId,
                    'category': match.category,
                    'message': match.message,
                    'suggestion': match.replacements[0] if match.replacements else None,
                    'offset': match.offset,
                    'length': match.errorLength,
                    'context': match.context,
                    'severity': self._get_severity(match.category)
                }
                issues.append(issue)
            
            # Calculate grammar score (0-100)
            score = max(0, 100 - (total_errors * 5))  # -5 points per error
            
            return {
                'score': score,
                'total_errors': total_errors,
                'issues': issues,
                'summary': self._generate_summary(issues)
            }
            
        except Exception as e:
            logger.error(f"Error analyzing text: {e}")
            return {
                'error': str(e),
                'score': 0,
                'issues': []
            }
    
    def _get_severity(self, category):
        """Determine severity level of an error"""
        severity_map = {
            'GRAMMAR': 'high',
            'SPELLING': 'medium',
            'STYLE': 'low',
            'PUNCTUATION': 'medium',
            'TYPOS': 'low'
        }
        return severity_map.get(category, 'medium')
    
    def _generate_summary(self, issues):
        """Generate a summary of grammar issues"""
        if not issues:
            return "No grammar issues found. Great job!"
        
        categories = {}
        for issue in issues:
            cat = issue['category']
            categories[cat] = categories.get(cat, 0) + 1
        
        summary_parts = []
        for cat, count in categories.items():
            summary_parts.append(f"{count} {cat.lower()} issue{'s' if count > 1 else ''}")
        
        return f"Found {len(issues)} total issues: {', '.join(summary_parts)}."
    
    def get_suggestions(self, text):
        """Get specific suggestions for improving the text"""
        if not self.tool:
            return []
        
        try:
            matches = self.tool.check(text)
            suggestions = []
            
            for match in matches:
                if match.replacements:
                    suggestion = {
                        'original': text[match.offset:match.offset + match.errorLength],
                        'suggestion': match.replacements[0],
                        'explanation': match.message,
                        'category': match.category
                    }
                    suggestions.append(suggestion)
            
            return suggestions
            
        except Exception as e:
            logger.error(f"Error getting suggestions: {e}")
            return []

# Initialize analyzer
analyzer = GrammarAnalyzer()

@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({
        'status': 'healthy',
        'service': 'grammar-checker',
        'tool_available': tool is not None
    })

@app.route('/analyze', methods=['POST'])
def analyze_essay():
    """Analyze essay for grammar issues"""
    try:
        data = request.get_json()
        
        if not data or 'text' not in data:
            return jsonify({'error': 'Text is required'}), 400
        
        text = data['text']
        
        if not text.strip():
            return jsonify({'error': 'Text cannot be empty'}), 400
        
        # Analyze the text
        result = analyzer.analyze_text(text)
        
        return jsonify(result)
        
    except Exception as e:
        logger.error(f"Error in analyze endpoint: {e}")
        return jsonify({'error': str(e)}), 500

@app.route('/suggestions', methods=['POST'])
def get_suggestions():
    """Get specific suggestions for text improvement"""
    try:
        data = request.get_json()
        
        if not data or 'text' not in data:
            return jsonify({'error': 'Text is required'}), 400
        
        text = data['text']
        suggestions = analyzer.get_suggestions(text)
        
        return jsonify({'suggestions': suggestions})
        
    except Exception as e:
        logger.error(f"Error in suggestions endpoint: {e}")
        return jsonify({'error': str(e)}), 500

@app.route('/check', methods=['POST'])
def quick_check():
    """Quick grammar check with basic info"""
    try:
        data = request.get_json()
        
        if not data or 'text' not in data:
            return jsonify({'error': 'Text is required'}), 400
        
        text = data['text']
        
        if not self.tool:
            return jsonify({'error': 'LanguageTool not available'}), 503
        
        matches = self.tool.check(text)
        
        return jsonify({
            'has_errors': len(matches) > 0,
            'error_count': len(matches),
            'score': max(0, 100 - (len(matches) * 5))
        })
        
    except Exception as e:
        logger.error(f"Error in quick check endpoint: {e}")
        return jsonify({'error': str(e)}), 500

if __name__ == '__main__':
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 5001
    logger.info(f"Starting Grammar Service on port {port}")
    app.run(host='0.0.0.0', port=port, debug=False) 

File: backend/src/services/grammarService.ts
import axios from 'axios';
import { logger } from '../utils/logger';

interface GrammarIssue {
  type: string;
  category: string;
  message: string;
  suggestion?: string;
  offset: number;
  length: number;
  context: string;
  severity: 'low' | 'medium' | 'high';
}

interface GrammarAnalysis {
  score: number;
  total_errors: number;
  issues: GrammarIssue[];
  summary: string;
  error?: string;
}

interface GrammarSuggestion {
  original: string;
  suggestion: string;
  explanation: string;
  category: string;
}

export class GrammarService {
  private baseUrl: string;
  private isAvailable: boolean = false;

  constructor() {
    this.baseUrl = process.env['GRAMMAR_SERVICE_URL'] || 'http://localhost:5001';
    this.checkHealth();
  }

  private async checkHealth(): Promise<void> {
    try {
      const response = await axios.get(`${this.baseUrl}/health`);
      this.isAvailable = response.data.tool_available;
      logger.info(`Grammar service health check: ${response.data.status}`);
    } catch (_error) {
      logger.warn('Grammar service not available, falling back to basic analysis');
      this.isAvailable = false;
    }
  }

  async analyzeEssay(text: string): Promise<GrammarAnalysis> {
    if (!this.isAvailable) return this.fallbackAnalysis(text);
    try {
      const response = await axios.post(`${this.baseUrl}/analyze`, { text });
      return response.data;
    } catch (error) {
      logger.error('Error calling grammar service:', error);
      return this.fallbackAnalysis(text);
    }
  }

  async getSuggestions(text: string): Promise<GrammarSuggestion[]> {
    if (!this.isAvailable) return [];
    try {
      const response = await axios.post(`${this.baseUrl}/suggestions`, { text });
      return response.data.suggestions || [];
    } catch (error) {
      logger.error('Error getting grammar suggestions:', error);
      return [];
    }
  }

  async quickCheck(text: string): Promise<{ has_errors: boolean; error_count: number; score: number; }> {
    if (!this.isAvailable) return this.fallbackQuickCheck(text);
    try {
      const response = await axios.post(`${this.baseUrl}/check`, { text });
      return response.data;
    } catch (error) {
      logger.error('Error in quick grammar check:', error);
      return this.fallbackQuickCheck(text);
    }
  }

  private fallbackAnalysis(text: string): GrammarAnalysis {
    const wordCount = text.split(/\s+/).length;
    const sentenceCount = text.split(/[.!?]+/).length - 1;
    const avgWordsPerSentence = wordCount / Math.max(sentenceCount, 1);
    const issues: GrammarIssue[] = [];
    let score = 100;
    if (avgWordsPerSentence > 25) {
      issues.push({
        type: 'LONG_SENTENCE',
        category: 'STYLE',
        message: 'Sentence is very long. Consider breaking it into shorter sentences.',
        offset: 0,
        length: text.length,
        context: text.substring(0, 100) + '...',
        severity: 'medium'
      });
      score -= 10;
    }
    if (wordCount < 50) {
      issues.push({
        type: 'SHORT_ESSAY',
        category: 'STYLE',
        message: 'Essay is very short. Consider adding more content.',
        offset: 0,
        length: text.length,
        context: text,
        severity: 'high'
      });
      score -= 20;
    }
    return {
      score: Math.max(0, score),
      total_errors: issues.length,
      issues,
      summary: issues.length > 0
        ? `Found ${issues.length} style issues. Consider using LanguageTool for detailed grammar analysis.`
        : 'Basic analysis complete. For detailed grammar checking, ensure LanguageTool service is running.'
    };
  }

  private fallbackQuickCheck(text: string): { has_errors: boolean; error_count: number; score: number; } {
    const wordCount = text.split(/\s+/).length;
    const hasErrors = wordCount < 50;
    return { has_errors: hasErrors, error_count: hasErrors ? 1 : 0, score: hasErrors ? 80 : 100 };
  }

  isServiceAvailable(): boolean {
    return this.isAvailable;
  }

  async refreshHealth(): Promise<void> {
    await this.checkHealth();
  }
}

export const grammarService = new GrammarService();

